File: uhrwerk-common/src/main/java/io/qimia/uhrwerk/common/model/StateModelBuilder.java
package io.qimia.uhrwerk.common.model;

import java.time.LocalDateTime;
import net.openhft.hashing.LongHashFunction;

public abstract class StateModelBuilder<B extends StateModelBuilder<B>> {
  LocalDateTime deactivatedTs;

  public B deactivatedTs(LocalDateTime deactivatedTs) {
    this.deactivatedTs = deactivatedTs;
    return getThis();
  }

  abstract B getThis();
}

File: uhrwerk-metastore/src/main/kotlin/io/qimia/uhrwerk/common/metastore/builders/StateModelBuilder.kt
package io.qimia.uhrwerk.common.metastore.builders

import java.time.LocalDateTime

abstract class StateModelBuilder<B : StateModelBuilder<B>?> {
    var deactivatedTs: LocalDateTime? = null
        protected set

    fun deactivatedTs(deactivatedTs: LocalDateTime?): B {
        this.deactivatedTs = deactivatedTs
        return getThis()
    }

    abstract fun getThis(): B

}